### PR TITLE
Jsdom

### PR DESCRIPTION
**Pull request by love-lena**
Added jsdom functionality

WIP

### CODE REVIEW

File: modules/jsdom/src/polyfills/glfw.ts [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+import {DOMWindow} from 'jsdom';
+import {Subscription} from 'rxjs';
+
+import {
+  isAltKey,
+  isCapsLock,
+  isCtrlKey,
+  isMetaKey,
+  isShiftKey
+} from '../../../glfw/src/events/event';
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:34]:
```suggestion
} from '@nvidia/glfw/events/event';
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+import GLFW, {GLFWStandardCursor} from '../../../glfw/src/glfw';
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:36]: Looks like we need to export the `GLFWStandardCursor` enum and `CURSOR_NORMAL`, `CURSOR_HIDDEN`, and `CURSOR_DISABLED` constants in the C++ addon, then we can do this:

```suggestion
import {GLFWCursorState, GLFWStandardCursor} from '@nvidia/glfw';
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+      value: 0,
+    },
+
+    altKey: {get: () => { return isAltKey(window.modifiers); }},
+    ctrlKey: {get: () => { return isCtrlKey(window.modifiers); }},
+    metaKey: {get: () => { return isMetaKey(window.modifiers); }},
+    shiftKey: {get: () => { return isShiftKey(window.modifiers); }},
+    capsLock: {get: () => { return isCapsLock(window.modifiers); }},
+
+    visable: {
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:129]:
```suggestion
    visible: {
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+      left: 0,
+      top: 0,
+      right: this.width,
+      bottom: this.height,
+    };
+  };
+
+  window.show = function() {
+    this.visible = true;
+    this._id || this._create();
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:268]:
```suggestion
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    glfw.showWindow(this._id);
+    return this;
+  };
+
+  window.hide = function() {
+    this.visible = false;
+    this._id || this._create();
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:275]:
```suggestion
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    glfw.hideWindow(this._id);
+    return this;
+  };
+
+  window.destoryGLFWWindow = function() {
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:280]:
```suggestion
  window.destroyGLFWWindow = function() {
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+          get height() { return self.height; },
+          set height(_: number) { self.height = _; },
+          get cursor() { return self.cursor; },
+          set cursor(_: any) { self.cursor = _; },
+        };
+      }
+    },
+
+    frameBufferWidth: {
+      get() { return this.width; },
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:216]:
```suggestion
      get() { return this._frameBufferWidth; },
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+    },
+    frameBufferHeight: {
+      get() { return this.height; },
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:219]:
```suggestion
      get() { return this._frameBufferHeight; },
```

File: modules/jsdom/src/polyfills/glfw.ts [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+  this._yscale            = event.yscale;
+  this._focused           = event.focused;
+  this._minimized         = event.minimized;
+  this._maximized         = event.maximized;
+  this._frameBufferWidth  = event.frameBufferWidth;
+  this._frameBufferHeight = event.frameBufferHeight;
+  this._devicePixelRatio =
+    Math.min(this._frameBufferWidth / this._width, this._frameBufferHeight / this._height);
+  dispatchGLFWEvent(this, event, this.Event);
+  if (event.type === 'close') { this._destroyGLFWWindow(); }
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/glfw.ts:388]:
```suggestion
  if (event.type === 'close') { this_destroyGLFWWindow(); }
```

Comment (love-lena) [reply modules/jsdom/src/polyfills/glfw.ts:388]: You corrected this to 
`this_destroyGLFWWindow()`
did you mean
`this.destroyGLFWWindow()`
?

File: modules/jsdom/src/polyfills/glfw.ts (hunk 2 above)
Comment (love-lena) [reply modules/jsdom/src/polyfills/glfw.ts:28]: It looks like GLFWStandardCursor is actually a const and not an enum?
glfw ln 295: `export const GLFWStandardCursor = {...`
not sure how this one should work

Comment (trxcllnt) [reply modules/jsdom/src/polyfills/glfw.ts:28]: in `glfw/src/index.ts` we need an `export { GLFWStandardCursor } from './glfw';`

File: modules/jsdom/src/polyfills/glfw.ts (hunk 9 above)
Comment (trxcllnt) [reply modules/jsdom/src/polyfills/glfw.ts]: Yep I typo'd the suggestion. Thought the method was still private with an underscore.

File: modules/jsdom/src/polyfills/events/monitor.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {map, publish, refCount} from 'rxjs/operators';
+
+import {glfw, GLFW, GLFWmonitor} from '../../../../glfw/src/glfw';
+import {Monitor} from '../../../../glfw/src/monitor';
```
Comment (trxcllnt) [inline modules/jsdom/src/polyfills/events/monitor.ts:18]: These paths won't be valid when we run the compiled code. We'll need to export these types from the glfw module in `index.ts`

**Comment by love-lena** [comment]
ofc i forget to remove the test commit

**Comment by love-lena** [comment]
Never worked on Lena's machine
  - playground
  - map-tile
  - interleaved-buffer
  - image-tile
  - icon
  - h3-grid
  - esri
  - ShanghaiData
  - 3d-tiles 

Breaks when moving to new jsdom
  - line
  - heatmap
  - bezier

Converted
  - 3d-heatmap
  - arc
  - geojson
  - highway
  - mesh
  - plot
  - point-cloud (with a new error)
  - scatterplot
  - scenegraph-layer (yeah but on and off errors with and without change)
  - screen-grid
  - tagmap
  - text-layer
  - trips
  - worldmap

**Pull request merged**